dockstore-webservice/src/main/resources/migrations.1.12.0.xml

```diff
@@ -248,4 +248,75 @@
             alter table version_metadata add constraint check_valid_doi check (doiurl like '10._%/_%' or doiurl is null);
         </sql>
     </changeSet>
+    <changeSet id="add_constraints_triggers_for_apptools" author="natalieperez">
+        <sql dbms="postgresql">
+            CREATE UNIQUE INDEX if not exists full_apptool_name ON apptool USING btree (sourcecontrol, organization, repository, workflowname) WHERE workflowname IS NOT NULL;
+            CREATE UNIQUE INDEX if not exists partial_apptool_name ON apptool USING btree (sourcecontrol, organization, repository) WHERE workflowname IS NULL;
+        </sql>
+        <createTable tableName="fullworkflowpath">
+            <column autoIncrement="true" name="id" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="fullworkflowpath_pkey"/>
+            </column>
+            <column name="organization" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="repository" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="sourcecontrol" type="TEXT">
+                <constraints nullable="false"/>
+            </column>
+            <column name="workflowname" type="TEXT"/>
+        </createTable>
+        <sql dbms="postgresql">
+            CREATE UNIQUE INDEX full_workflow_name_masterlist ON fullworkflowpath USING btree (sourcecontrol, organization, repository, workflowname) WHERE workflowname IS NOT NULL;
+            CREATE UNIQUE INDEX partial_workflow_name_masterlist ON fullworkflowpath USING btree (sourcecontrol, organization, repository) WHERE workflowname IS NULL;
+
+            CREATE OR REPLACE FUNCTION workflow_path_insert_trigger_fnc()
+            RETURNS trigger AS
```

Comment:
A bit of a summary of my thoughts essentially when I saw trigger vs JPA lifecycle. 
I guess this could make more sense if we go with multiple instances of the webservice.

Reply:
Oops, forgot the link https://stackoverflow.com/questions/5249004/triggers-vs-jpa-prepersist-for-creation-and-update-timestamps-pros-and-cons